ls1intum/Artemis
# Simplify Compass

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest2.ase.in.tum.de.

### Motivation and Context
The compass was using too many local maps and was causing too much data transfer when keeping these maps in Hazelcast. Therefore we needed a solution that doesn't need Hazelcast and all these maps but can work on multi node setups.

### Description
This PR changes the whole pipeline of Compass. It removes all the parts of Compass and replaces them with new and simpler classes that uses database. Before the feedback suggestion calculation was done for all the submission at once and repeated whenever tutor adds a new assessment. However now, Compass creates clusters and elements in database once the due date is over with a scheduled operation. With them Compass can suggest feedback when tutor wants it by selecting feedback from the feedbacks of the same cluster for each element.

### Steps for Testing

1. Log in to Artemis
2. Create a modeling exercise with Compass enabled
3. Set a due date for the exercise
4. Create several similar submissions until the due date
5. Wait until due date is passed
6. Start assessing modeling exercises
7. There should be suggestions for the next exercises after you submit a manual assessment for the first one

### Test Coverage
ModelClusterFactory.java 91%
FeedbackSelector.java 95%
CompassService.java 79%

### Screenshots
#### 1. Create the exercise with compass enabled
https://user-images.githubusercontent.com/16825065/121805412-79064680-cc4b-11eb-96f0-d572e59495ba.mp4
#### 2. Submit other similar submissions
#### 3. Update due date of the exercise and wait until it has passed
https://user-images.githubusercontent.com/16825065/121805416-802d5480-cc4b-11eb-953b-e8dee2aa2c8c.mp4
#### 4. Assess the submissions and see the automatic assessment suggestions
https://user-images.githubusercontent.com/16825065/121805424-86bbcc00-cc4b-11eb-9838-7d97ba6a825d.mp4
